dm/pkg/checker/table_structure.go

```diff
@@ -60,71 +74,128 @@ func (o *incompatibilityOption) String() string {
 type TablesChecker struct {
 	db     *sql.DB
 	dbinfo *dbutil.DBConfig
-	tables map[string][]string // schema => []table; if []table is empty, query tables from db
+	tables []*filter.Table // tableID => filter.Table
 }
 
 // NewTablesChecker returns a RealChecker.
-func NewTablesChecker(db *sql.DB, dbinfo *dbutil.DBConfig, tables map[string][]string) RealChecker {
-	return &TablesChecker{
+func NewTablesChecker(db *sql.DB, dbinfo *dbutil.DBConfig, tables []*filter.Table) RealChecker {
+	c := &TablesChecker{
 		db:     db,
 		dbinfo: dbinfo,
 		tables: tables,
 	}
+	return c
 }
 
 // Check implements RealChecker interface.
-func (c *TablesChecker) Check(ctx context.Context) *Result {
+func (c *TablesChecker) Check(ctx context.Context) (*Result, error) {
 	r := &Result{
 		Name:  c.Name(),
 		Desc:  "check compatibility of table structure",
 		State: StateSuccess,
 		Extra: fmt.Sprintf("address of db instance - %s:%d", c.dbinfo.Host, c.dbinfo.Port),
 	}
 
+	startTime := time.Now()
+
+	concurrency := len(c.tables)/tablesForOneThread + 1
+	if concurrency > maxThreadNum {
+		concurrency = maxThreadNum
+	}
+	maxConnections, err := utils.GetMaxConnections(ctx, c.db)
+	if err != nil {
+		return r, err
+	}
+	if concurrency > maxConnections {
+		concurrency = maxConnections
```

Comment:
```suggestion
		concurrency = maxConnections / 2
```
as @GMHDBJD  suggestions, we better not use all connections 😬